prov/rxd/src/rxd_cq.c

```diff
@@ -255,21 +255,24 @@ void rxd_post_cts(struct rxd_ep *rxd_ep, struct rxd_x_entry *rx_entry,
 	struct rxd_pkt_entry *pkt_entry;
 	int ret;
 
-	rx_entry->peer = rts_pkt->peer;
-	rx_entry->window = rts_pkt->pkt->ctrl.window;
-	rx_entry->key = rts_pkt->pkt->hdr.key;
-	rx_entry->tx_id = rts_pkt->pkt->hdr.tx_id;
-	rx_entry->seg_size = rts_pkt->pkt->ctrl.seg_size;
-	rx_entry->state = RXD_ACK;
-	rx_entry->next_start = rx_entry->window;
-	rx_entry->num_segs = rxd_div_ceil(rts_pkt->pkt->ctrl.size,
-					  rts_pkt->pkt->ctrl.seg_size);
-
-	if (rts_pkt->pkt->hdr.flags & RXD_REMOTE_CQ_DATA) {
-		rx_entry->cq_entry.flags |= FI_REMOTE_CQ_DATA;
-		rx_entry->cq_entry.data = rts_pkt->pkt->ctrl.data;
+	if (rx_entry->state == RXD_RTS) {
+		rx_entry->peer = rts_pkt->peer;
+		rx_entry->key = rts_pkt->pkt->hdr.key;
+		rx_entry->tx_id = rts_pkt->pkt->hdr.tx_id;
+		rx_entry->seg_size = rts_pkt->pkt->ctrl.seg_size;
+		rx_entry->num_segs = rxd_div_ceil(rts_pkt->pkt->ctrl.size,
+						  rts_pkt->pkt->ctrl.seg_size);
+		rx_entry->window = rts_pkt->pkt->ctrl.window;
+		rx_entry->next_start = rx_entry->window;
+
+		if (rts_pkt->pkt->hdr.flags & RXD_REMOTE_CQ_DATA) {
+			rx_entry->cq_entry.flags |= FI_REMOTE_CQ_DATA;
+			rx_entry->cq_entry.data = rts_pkt->pkt->ctrl.data;
+		}
+		rx_entry->cq_entry.len = rts_pkt->pkt->ctrl.size;
+
+		rx_entry->state = RXD_ACK;
```

Comment:
This initialization looks like it would be better placed where the rx_entry is allocated than as part of the post cts call.

Reply:
This message information isn't actually known until the RTS message is received, so it can't be initialized when the rx_entry is actually posted.